File: evaluation/S0/cpp/Hard/2205-height-of-binary-tree-after-subtree-removal-queries.cpp
// Source: https://leetcode.com/problems/height-of-binary-tree-after-subtree-removal-queries/   |   Difficulty: Hard
//
// Problem Description:
// You are given the root of a binary tree with n nodes. Each node is assigned a unique value from 1 to n. You are also given an array queries of size m.
//
// You have to perform m independent queries on the tree where in the ith query you do the following:
//
//
// 	Remove the subtree rooted at the node with the value queries[i] from the tree. It is guaranteed that queries[i] will not be equal to the value of the root.
//
//
// Return an array answer of size m where answer[i] is the height of the tree after performing the ith query.
//
// Note:
//
//
// 	The queries are independent, so the tree returns to its initial state after each query.
// 	The height of a tree is the number of edges in the longest simple path from the root to some node in the tree.
//
// Example:
// Input: root = [1,3,4,2,null,6,5,null,null,null,null,null,7], queries = [4]
// Output: [2]
// Explanation: The diagram above shows the tree after removing the subtree rooted at node with value 4.
// The height of the tree is 2 (The path 1 -> 3 -> 2).
//
// Constraints:
// The number of nodes in the tree is n.
// 	2 <= n <= 105
// 	1 <= Node.val <= n
// 	All the values in the tree are unique.
// 	m == queries.length
// 	1 <= m <= min(n, 104)
// 	1 <= queries[i] <= n
// 	queries[i] != root.val
//

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> treeQueries(TreeNode* root, vector<int>& queries) {
        int n = 0;
        unordered_map<int, vector<int>> adj;
        unordered_map<int, int> parent;
        unordered_map<int, int> depth;
        unordered_map<int, int> height;
        
        // Build tree as adjacency list and record parent and depth
        function<void(TreeNode*, int, int)> dfs = [&](TreeNode* node, int par, int d) {
            if (!node) return;
            n++;
            depth[node->val] = d;
            parent[node->val] = par;
            if (par != -1) {
                adj[par].push_back(node->val);
            }
            dfs(node->left, node->val, d + 1);
            dfs(node->right, node->val, d + 1);
        };
        
        dfs(root, -1, 0);
        
        // Compute height of each subtree
        function<int(int)> computeHeight = [&](int u) {
            int h = 0;
            for (int v : adj[u]) {
                h = max(h, computeHeight(v) + 1);
            }
            height[u] = h;
            return h;
        };
        
        computeHeight(root->val);
        
        // Precompute the maximum heights excluding each child
        unordered_map<int, vector<int>> maxExclChild;
        
        function<void(int)> precomputeMaxExcl = [&](int u) {
            vector<int> childHeights;
            for (int v : adj[u]) {
                childHeights.push_back(height[v]);
            }
            
            int m = childHeights.size();
            vector<int> prefixMax(m + 1, -1), suffixMax(m + 1, -1);
            
            for (int i = 0; i < m; i++) {
                prefixMax[i + 1] = max(prefixMax[i], childHeights[i]);
            }
            for (int i = m - 1; i >= 0; i--) {
                suffixMax[i] = max(suffixMax[i + 1], childHeights[i]);
            }
            
            for (int i = 0; i < m; i++) {
                maxExclChild[u].push_back(max(prefixMax[i], suffixMax[i + 1]));
            }
            
            for (int v : adj[u]) {
                precomputeMaxExcl(v);
            }
        };   
        precomputeMaxExcl(root->val);
        vector<int> result;
        for (int q : queries) {
            int par = parent[q];
            if (par == -1) {
                result.push_back(0);
                continue;
            }
            int idx = -1;
            for (int i = 0; i < adj[par].size(); i++) {
                if (adj[par][i] == q) {
                    idx = i;
                    break;
                }
            }
            int exclHeight = maxExclChild[par][idx];
            int newHeight = depth[par] + (exclHeight == -1 ? 0 : exclHeight + 1);
            result.push_back(newHeight);
        }
        return result;       
    }
};

File: evaluation/S3/python/Hard/2205-height-of-binary-tree-after-subtree-removal-queries.py
# Source: https://leetcode.com/problems/height-of-binary-tree-after-subtree-removal-queries/   |   Difficulty: Hard
#
# Problem Description:
# You are given the root of a binary tree with n nodes. Each node is assigned a unique value from 1 to n. You are also given an array queries of size m.
#
# You have to perform m independent queries on the tree where in the ith query you do the following:
#
#
# 	Remove the subtree rooted at the node with the value queries[i] from the tree. It is guaranteed that queries[i] will not be equal to the value of the root.
#
#
# Return an array answer of size m where answer[i] is the height of the tree after performing the ith query.
#
# Note:
#
#
# 	The queries are independent, so the tree returns to its initial state after each query.
# 	The height of a tree is the number of edges in the longest simple path from the root to some node in the tree.
#
# Example:
# Input: root = [1,3,4,2,null,6,5,null,null,null,null,null,7], queries = [4]
# Output: [2]
# Explanation: The diagram above shows the tree after removing the subtree rooted at node with value 4.
# The height of the tree is 2 (The path 1 -> 3 -> 2).
#
# Constraints:
# The number of nodes in the tree is n.
# 	2 <= n <= 105
# 	1 <= Node.val <= n
# 	All the values in the tree are unique.
# 	m == queries.length
# 	1 <= m <= min(n, 104)
# 	1 <= queries[i] <= n
# 	queries[i] != root.val
#
# Helpful references (internal KB):
# - Lowest Common Ancestor (tree, array, dfs)
#   • When to use: Use this algorithm to efficiently find the lowest common ancestor of two nodes in a tree. It is particularly effective for answering multiple LCA queries on a static tree after initial preprocessing.
#   • Idea: This algorithm preprocesses a tree using a DFS to build an Euler tour and record node depths. It transforms Lowest Common Ancestor (LCA) queries into Range Minimum Queries (RMQ) on the Euler tour, enabling O(1) query time after O(N log N) preprocessing.
#   • Invariants: During DFS, `euler` contains the sequence of visited nodes, and `height` stores each node's depth from the root.; The `first` array correctly maps each node to its initial index in the `euler` tour.
#   • Tips: Perform a single DFS to compute node depths and construct the Euler tour along with first occurrences.; Utilize a Sparse Table or Segment Tree for O(1) or O(log N) Range Minimum Query after preprocessing.
#   • Pitfalls: Incorrectly building the Euler tour, especially missing the 'return' visits to parent nodes.; Off-by-one errors when defining the range for the Range Minimum Query on the Euler tour.
# - Lowest Common Ancestor - Binary Lifting (tree, array, binary-lifting, dfs)
#   • When to use: Use binary lifting when you need to answer multiple Lowest Common Ancestor (LCA) queries or ancestor relationship checks efficiently on a static tree.
#   • Idea: Binary lifting precomputes ancestor pointers using DFS to answer Lowest Common Ancestor (LCA) queries and ancestor checks in O(log N) time per query, after O(N log N) preprocessing.
#   • Invariants: For any node `u` and power `k`, `up[u][k]` is the 2^k-th ancestor of `u`.; `depth[u]` correctly represents the distance from the root to node `u`.
#   • Tips: Precompute `up[u][k]` as the 2^k-th ancestor of `u`.; Use DFS to fill the `up` array and node depths.
#   • Pitfalls: Incorrectly calculating `log N` for array dimensions.; Failing to handle the root node's ancestors (e.g., `up[root][k]` should be itself or an invalid node).
# - Paint the edges of the tree (tree, array, dfs, lca, segment-tree)
#   • When to use: Used for efficiently handling queries on a tree that involve updating edge properties (like color) and querying aggregated properties (like count of colored edges) along paths between two nodes.
#   • Idea: This technique uses an Euler tour and two segment trees to maintain edge properties and query path sums. It allows for O(log N) updates and queries after an O(N) preprocessing step.
#   • Invariants: The Euler tour correctly records each edge traversal exactly twice (once forward, once backward).; Each segment tree accurately maintains the sum of colored edges for its respective traversal direction.
#   • Tips: Reduce general path queries (i, j) to ancestor-descendant queries (LCA(i,j), i) and (LCA(i,j), j) using Lowest Common Ancestor.; The Euler tour transforms tree paths into contiguous segments in an array, enabling segment tree operations.
#   • Pitfalls: Incorrectly handling the LCA step can lead to wrong path decomposition for general queries.; Off-by-one errors when defining the segment tree query range [p, q-1] in the Euler tour.
#
# Notes:
# - Keep the given function signature unchanged.
# - Use only the Python standard library; no third-party packages.
# - Do not change input/output formats or return value semantics as defined by the template/problem.
# - Conformance: use features compatible with LeetCode’s Python3 runtime; do NOT rely on undefined/implementation-specific behavior.
# - If you introduce any helper functions/classes, include their full implementations in this file; no hidden helpers.

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def treeQueries(self, root, queries):
        """
        :type root: Optional[TreeNode]
        :type queries: List[int]
        :rtype: List[int]
        """
        from collections import defaultdict

        height = defaultdict(int)

        def dfs(node, h):
            if not node:
                return h - 1
            height[node.val] = max(height[node.val], h)
            dfs(node.left, h + 1)
            dfs(node.right, h + 1)

        dfs(root, 0)

        max_height = max(height.values())
        count = [0] * (max_height + 1)
        for h in height.values():
            count[h] += 1

        result = []
        for q in queries:
            h = height[q]
            if count[h] == 1:
                result.append(h - 1)
            else:
                result.append(max_height)
        return result